### PR TITLE
Feature/dataframe to string

### PR DESCRIPTION
**Pull request by tadeohepperle**
Override toString() method of DataFrame to make it easier to assess the content and dimensions of DataFrames by just printing them to the console.
````dart
print(dataFrame)
````
will yield:

````txt
DataFrame (13 x 10)
 id   age   salary   children   gender           profession   ...   weight
  1    25    30000          2        M              Teacher   ...     78.3
  2    46    85000          0        M              Manager   ...     45.2
  3    36    45000          1        F              Teacher   ...     98.4
  4    23    10000          5        M   Mushroom Collector   ...     57.4
  5    22    30000          2        M              Butcher   ...     87.9
...   ...      ...        ...      ...                  ...   ...      ...
  9    23      N/A          2        M           Unemployed   ...     56.7
 10    25    32000          4        F              Teacher   ...     98.7
 11    49    34700          0        M              Plumber   ...    120.3
 12    36    45000          1        F            Paramedic   ...     67.9
 13    23    42900          2        M           Researcher   ...     92.3


````

### CODE REVIEW

File: lib/src/data_frame/data_frame_impl.dart [hunk 1]
```diff
@@ -185,14 +182,26 @@ class DataFrameImpl with SerializableMixin implements DataFrame {
...
+  @override
+  String toString({int maxRows = 10, int maxCols = 7}) {
+    return dataFrameToString(
+      this,
+      maxRows: maxRows,
+      maxCols: maxCols,
+    );
+  }
+
+  String visua() {
```
Comment (gyrdym) [inline lib/src/data_frame/data_frame_impl.dart]: @tadeohepperle apparently, you forgot to remove this method. Could you please remove it?

Comment (tadeohepperle) [reply lib/src/data_frame/data_frame_impl.dart]: @gyrdym Ah yes I did. Sorry.

File: test/data_frame/data_frame_to_string_test.dart [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ 6    28    82000          0        F            Scientist                   10      true      179     98.3
+ 7    46    85000          0        M            Scientist                    7     false      176     67.8
+ 8    36    45000          1        F              Teacher                    4      true      165     76.8
+ 9    23      N/A          2        M           Unemployed                    0      true      179     56.7
+10    25    32000          4        F              Teacher                    5      true      189     98.7
+11    49    34700          0        M              Plumber                    7      true      176    120.3
+12    36    45000          1        F            Paramedic                    4      true      165     67.9
+13    23    42900          2        M           Researcher                    0      true      179     92.3''';
+
+    test('should print dataframe correctly with default maxRows=10, maxCols=7',
```
Comment (gyrdym) [inline test/data_frame/data_frame_to_string_test.dart]: @tadeohepperle could you please add tests for edge cases? E.g. `maxRows = 0`, `maxRows=1`, `maxRows=2`, and the same cases for `maxCols`?

File: lib/src/data_frame/helpers/data_frame_to_string.dart [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+const kPadding = 3;
+const kSkippingSymbol = '...';
+
+String dataFrameToString(DataFrame dataFrame, {maxRows = 10, maxCols = 7}) {
+  final rowCountStartEnd = [
+    (maxRows / 2).toInt(),
+    maxRows - (maxRows / 2).toInt()
+  ];
```
Comment (gyrdym) [inline lib/src/data_frame/helpers/data_frame_to_string.dart]: @tadeohepperle could you please add a new line after this line (for better formatting)?

File: lib/src/data_frame/helpers/data_frame_to_string.dart [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  maxCols =
+      dataFrame.header.length < maxCols ? dataFrame.header.length : maxCols;
+  maxRows = dataFrame.rows.length < maxRows ? dataFrame.rows.length : maxRows;
```
Comment (gyrdym) [inline lib/src/data_frame/helpers/data_frame_to_string.dart]: @tadeohepperle could you please add a new line after this line (for better formatting)?

File: lib/src/data_frame/helpers/data_frame_to_string.dart [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    if (i < rowCountStartEnd[0] || i >= nRows - rowCountStartEnd[1]) {
+      var j = 0;
+      var seriesCounter = 0;
+      for (var value in row) {
+        if (j < colCountStartEnd[0] || j >= nCols - colCountStartEnd[1]) {
+          var displayData = seriesDisplayData[seriesCounter];
+          var valueString = value.toString();
+          displayData.data.add(valueString);
+          if (displayData.maxChars < valueString.length) {
+            displayData.maxChars = valueString.length;
```
Comment (gyrdym) [inline lib/src/data_frame/helpers/data_frame_to_string.dart]: @tadeohepperle you can use 

```dart
displayData.maxChars = math.max(displayData.maxChars, valueString.length)
```

instead of `if` operator.

You should import `math` module before:

`import 'dart:math' as math;`

File: lib/src/data_frame/helpers/data_frame_to_string.dart (hunk 4 above)
Comment (gyrdym) [reply lib/src/data_frame/helpers/data_frame_to_string.dart]: @tadeohepperle also, you can use 

```dart
maxCols = math.max(dataFrame.header.length, maxCols);
maxCols = math.max(dataFrame.rows.length, maxRows);
```

You should import math module before:

import 'dart:math' as math;

File: lib/src/data_frame/helpers/data_frame_to_string.dart [hunk 6]
```diff
@@ -0,0 +1,87 @@
+import 'package:ml_dataframe/ml_dataframe.dart';
```
Comment (gyrdym) [inline lib/src/data_frame/helpers/data_frame_to_string.dart]: @tadeohepperle could you please replace the index file import with the direct one? I see, that you use `DataFrame` class, so you can do the following import:

```dart
import 'package:ml_dataframe/src/data_frame/data_frame.dart';
```

instead of 

```dart
import 'package:ml_dataframe/ml_dataframe.dart';
```

Index files such as `package:ml_dataframe/ml_dataframe.dart` are only for library external usage, inside the library, it is better to use  imports from the `src` directory

File: lib/src/data_frame/helpers/data_frame_to_string.dart [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        } else if (j == colCountStartEnd[0]) {
+          seriesDisplayData[seriesCounter].data.add(kSkippingSymbol);
+          seriesDisplayData[seriesCounter].maxChars = kSkippingSymbol.length;
+          seriesCounter++;
+        }
+        j++;
+      }
+    } else if (i == rowCountStartEnd[0]) {
+      for (var d in seriesDisplayData) {
+        d.maxChars = d.maxChars < kSkippingSymbol.length
```
Comment (gyrdym) [inline lib/src/data_frame/helpers/data_frame_to_string.dart]: @tadeohepperle you can also use the same trick with `math.max` here

**Comment by gyrdym** [comment]
@tadeohepperle could you please update the version in pubspec.yaml and add a changelog record? I think it's better to update the minor version (0.5.1 -> 0.6.0)

File: test/data_frame/data_frame_to_string_test.dart (hunk 2 above)
Comment (tadeohepperle) [reply test/data_frame/data_frame_to_string_test.dart]: @gyrdym Currently some edge cases are not working well, you are right, I should have thought about this. I'll fix.

**Review by gyrdym — APPROVED** [review]
Looks good.
